### ISSUE TITLE
Indexing into numbers should be an error

### ISSUE THREAD

**Issue by danvk** [Microsoft/TypeScript#15682]
I was in a situation where I moved from `number[][]` to `number[]` and was surprised that some errors along these lines didn't get flagged.

**TypeScript Version:**  2.2.1

**Code**

```ts
const x = 10;  // x's type is number
const y = x[12];  // y's type is any. This is not an error.
```

**Expected behavior:**

Indexing into a number should be an error.

**Actual behavior:**

Indexing into a number is allowed, unless you set `noImplicitAny`.

**Comment by mhegazy** [comment]
this looks more like a lint rule to be frank. the compiler already flags such indexes as errors under `--noImplicitAny`, going further and classifying the error based on the target of the index expression would be in the linting domain.

**Comment by danvk** [comment]
Is there any scenario where that would be valid code?

**Comment by mhegazy** [comment]
But so is:

```ts
var x = {};
x[2];

function f() {}
f["bar"];
```

and these are all fagged as errors under `--noImplicitAny`.

**Comment by mhegazy** [comment]
I should add that we recommend always using `--noImplicitAny` for TS code.

**Comment by danvk** [comment]
Those feel different to me since they could become valid at some point:

```
> x = {}
Object {}
> x[10] = true
true
> x[10]
true
```

vs

```
> x = 2
2
> x[10] = true
true
> x[10]
undefined
```

It's just surprising to me that `number[]` and `number[][]` aren't really distinguishable without `noImplicitAny`.

**Comment by mhegazy** [comment]
> Those feel different to me since they could become valid at some point

and that is why i am saying this should be done by a linter rule instead.

**Comment by danvk** [comment]
OK, I'll close out. I guess another reason to always set `noImplicitAny`.
